In [1]:
!pip install pandas
!pip install numpy

In [1]:
### Importing the dependencies ###
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

In [2]:
#### Loading the dataset using pandas library ####

df = pd.read_csv(r'C:\Users\RL465FV\OneDrive - EY\Desktop\My_Python\Credit Card Project\creditcard.csv')
df.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

Variables/Columns present here mean-

Time: Here it is the time elapsed in seconds from the time elapsed from after the first transaction which is the first row/record here. For eg: If 10th transaction has 60 as time that means the 10th transcation happened 60 seconds after the first transaction (top row/record).

Columns (V1- V28) - these are the details about the transactions and credit card which were sensitive so the data provider converted these into numeric values using PCA(Principal Component Analysis)

Amount - Its in USD. 

Class - It has 0- Genuine/Legit transaction and 1- Fraudulent tranasaction. 

In [3]:
### Bottom of the dataset ###
df.tail(10)

Time         V1         V2        V3        V4        V5  \
284797  172782.0  -0.241923   0.712247  0.399806 -0.463406  0.244531   
284798  172782.0   0.219529   0.881246 -0.635891  0.960928 -0.152971   
284799  172783.0  -1.775135  -0.004235  1.189786  0.331096  1.196063   
284800  172784.0   2.039560  -0.175233 -1.196825  0.234580 -0.008713   
284801  172785.0   0.120316   0.931005 -0.546012 -0.745097  1.130314   
284802  172786.0 -11.881118  10.071785 -9.834783 -2.066656 -5.364473   
284803  172787.0  -0.732789  -0.055080  2.035030 -0.738589  0.868229   
284804  172788.0   1.919565  -0.301254 -3.249640 -0.557828  2.630515   
284805  172788.0  -0.240440   0.530483  0.702510  0.689799 -0.377961   
284806  172792.0  -0.533413  -0.189733  0.703337 -0.506271 -0.012546   

              V6        V7        V8        V9  ...       V21       V22  \
284797 -1.343668  0.929369 -0.206210  0.106234  ... -0.228876 -0.514376   
284798 -1.014307  0.427126  0.121340 -0.285670  ...  0.099936  0.337120   
284799  5.519980 -1.518185  2.080825  1.159498  ...  0.103302  0.654850   
284800 -0.726571  0.017050 -0.118228  0.435402  ... -0.268048 -0.717211   
284801 -0.235973  0.812722  0.115093 -0.204064  ... -0.314205 -0.808520   
284802 -2.606837 -4.918215  7.305334  1.914428  ...  0.213454  0.111864   
284803  1.058415  0.024330  0.294869  0.584800  ...  0.214205  0.924384   
284804  3.031260 -0.296827  0.708417  0.432454  ...  0.232045  0.578229   
284805  0.623708 -0.686180  0.679145  0.392087  ...  0.265245  0.800049   
284806 -0.649617  1.577006 -0.414650  0.486180  ...  0.261057  0.643078   

             V23       V24       V25       V26       V27       V28  Amount  \
284797  0.279598  0.371441 -0.559238  0.113144  0.131507  0.081265    5.49   
284798  0.251791  0.057688 -1.508368  0.144023  0.181205  0.215243   24.05   
284799 -0.348929  0.745323  0.704545 -0.127579  0.454379  0.130308   79.99   
284800  0.297930 -0.359769 -0.315610  0.201114 -0.080826 -0.075071    2.68   
284801  0.050343  0.102800 -0.435870  0.124079  0.217940  0.068803    2.69   
284802  1.014480 -0.509348  1.436807  0.250034  0.943651  0.823731    0.77   
284803  0.012463 -1.016226 -0.606624 -0.395255  0.068472 -0.053527   24.79   
284804 -0.037501  0.640134  0.265745 -0.087371  0.004455 -0.026561   67.88   
284805 -0.163298  0.123205 -0.569159  0.546668  0.108821  0.104533   10.00   
284806  0.376777  0.008797 -0.473649 -0.818267 -0.002415  0.013649  217.00   

        Class  
284797      0  
284798      0  
284799      0  
284800      0  
284801      0  
284802      0  
284803      0  
284804      0  
284805      0  
284806      0  

[10 rows x 31 columns]

Here the time of last transaction is 172792 seconds, which means it has trnasactnion of 47.999777 hours i.e approx 2 days.

In [4]:
### Dataset information ###
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

In [5]:
### Another way of Checking the number of missing values in each column ###
df.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

In [6]:
### Distribution of legit and fraudelent transactions ###
df['Class'].value_counts()

0    284315
1       492
Name: Class, dtype: int64

### The data is highly imbalanced
This data is imbalanced because more than 99% of data is present in one class (0) i.e. normal transactions. Thus, we cannot feed this data and train the machine learning model because it will not be able to recognise the fraudelent data as the data point is very low in this class (1).

In [7]:
## Separating the data based on Legit (0) and Fraudelent(1) transaction for data analysis ##
legit = df[df.Class == 0]
fraud = df[df.Class == 1]
legit.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

In [8]:
print(legit.shape)
print(fraud.shape)

(284315, 31)
(492, 31)


In [9]:
### Statistical Measures of the data ###
legit.Amount.describe()

count    284315.000000
mean         88.291022
std         250.105092
min           0.000000
25%           5.650000
50%          22.000000
75%          77.050000
max       25691.160000
Name: Amount, dtype: float64

Here, percentile means
25% of the data is below $5.65

In [10]:
fraud.Amount.describe()

count     492.000000
mean      122.211321
std       256.683288
min         0.000000
25%         1.000000
50%         9.250000
75%       105.890000
max      2125.870000
Name: Amount, dtype: float64

In [11]:
## Comparing the values for both the transactions ##
df.groupby(df.Class).mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      94838.202258  0.008258 -0.006271  0.012171 -0.007860  0.005453   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0      0.002419  0.009637 -0.000987  0.004467  ... -0.000644 -0.001235   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0     -0.000024  0.000070  0.000182 -0.000072 -0.000089 -0.000295 -0.000131   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       88.291022  
1      122.211321  

[2 rows x 30 columns]

Under-sampling:
Since it is highly imbalanced data and is unfit to go ahead. So here using Under-Sampling technique to correct the imbalanced data.


Now building a sample data from the original data for Normal/Legit and Fraudelent txns.

No of Fraudelent txns - 492

No of Normal/Legit txns - 284315

So, now from these 284315 normal txns 492 random txns will be taken out and then the ratio of fraudelent and normal txns will be 50:50 which will be better for training the machine learning model.

In [12]:
### Getting 492 random records from the Legit records ###
legit_sample = legit.sample(n=492)
legit_sample.shape

(492, 31)

In [13]:
## Concatenating the fraud and legit_sample tables ###

new_df = pd.concat([legit_sample, fraud], axis = 0)
new_df.shape

(984, 31)

Here, axis = 0 means rows and axis = 1 means columns. So, here the the records will get added row-wise 1 by 1 and not column-wise.

In [14]:
new_df.head()

Time        V1        V2        V3        V4        V5        V6  \
174305  121859.0 -0.346410  0.398823 -2.212968 -2.051168  0.598982 -2.358272   
269283  163614.0  2.056941  0.115745 -1.703518  0.434246  0.369519 -0.902841   
78613    57614.0  1.203190  0.056183  0.576976  0.572712 -0.708321 -0.835329   
71973    54515.0 -0.205346  1.065287  1.166512  1.501682 -0.384694 -0.340569   
170113  120023.0  2.038026 -0.144458 -1.219570  0.202584  0.110969 -0.583490   

              V7        V8        V9  ...       V21       V22       V23  \
174305  0.901844 -0.197581 -1.206912  ...  0.673596  2.029161 -0.162931   
269283  0.180290 -0.226781  0.528093  ... -0.360418 -0.963085  0.339668   
78613  -0.180014  0.047664  0.088710  ... -0.216942 -0.765436  0.175240   
71973   1.055168 -0.295705 -0.824004  ...  0.133631  0.483372 -0.119879   
170113  0.036896 -0.118135  0.356666  ... -0.261527 -0.663087  0.288902   

             V24       V25       V26       V27       V28  Amount  Class  
174305  0.130769 -0.968043 -0.127059  0.549542  0.291011   20.00      0  
269283  0.561755 -0.273752  0.169598 -0.065790 -0.029244    9.84      0  
78613   0.472940  0.089510  0.071583 -0.046134  0.006256    0.89      0  
71973   0.432801  0.033032 -0.135515  0.008659  0.032113  101.90      0  
170113 -0.474942 -0.301338  0.209406 -0.074052 -0.075384    1.98      0  

[5 rows x 31 columns]

In [15]:
new_df.tail()

Time        V1        V2        V3        V4        V5        V6  \
279863  169142.0 -1.927883  1.125653 -4.518331  1.749293 -1.566487 -2.010494   
280143  169347.0  1.378559  1.289381 -5.004247  1.411850  0.442581 -1.326536   
280149  169351.0 -0.676143  1.126366 -2.213700  0.468308 -1.120541 -0.003346   
281144  169966.0 -3.113832  0.585864 -5.399730  1.817092 -0.840618 -2.943548   
281674  170348.0  1.991976  0.158476 -2.583441  0.408670  1.151147 -0.096695   

              V7        V8        V9  ...       V21       V22       V23  \
279863 -0.882850  0.697211 -2.064945  ...  0.778584 -0.319189  0.639419   
280143 -1.413170  0.248525 -1.127396  ...  0.370612  0.028234 -0.145640   
280149 -2.234739  1.210158 -0.652250  ...  0.751826  0.834108  0.190944   
281144 -2.208002  1.058733 -1.632333  ...  0.583276 -0.269209 -0.456108   
281674  0.223050 -0.068384  0.577829  ... -0.164350 -0.295135 -0.072173   

             V24       V25       V26       V27       V28  Amount  Class  
279863 -0.294885  0.537503  0.788395  0.292680  0.147968  390.00      1  
280143 -0.081049  0.521875  0.739467  0.389152  0.186637    0.76      1  
280149  0.032070 -0.739695  0.471111  0.385107  0.194361   77.89      1  
281144 -0.183659 -0.328168  0.606116  0.884876 -0.253700  245.00      1  
281674 -0.450261  0.313267 -0.289617  0.002988 -0.015309   42.53      1  

[5 rows x 31 columns]

So here as legit_sample was put first into concat hence that is the one on top of this table (new_df).

In [16]:
new_df['Class'].value_counts()

0    492
1    492
Name: Class, dtype: int64

So here we can see now our data is uniformly distributed i.e. 492 txns each for legit and fraudulent class.

In [17]:
new_df.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      93948.042683  0.068193 -0.012611 -0.018253  0.052782 -0.005316   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0      0.030328 -0.092363  0.050981 -0.014175  ... -0.034923  0.018643   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23      V24       V25       V26       V27       V28  \
Class                                                                        
0     -0.047165 -0.000099 -0.02390 -0.020395  0.003385 -0.014201  0.003263   
1      0.014049 -0.040308 -0.10513  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       81.260732  
1      122.211321  

[2 rows x 30 columns]

Here after the comparing the mean of the legit txns the values are very much similar. This means the property/characteristic of the legit txns is same which will help the machine learning model to differentiate between the legit and fraud txn.

This step is important which helps us check whther the samples pulled are right for going ahead or not.

In [18]:
### Splitting the data into features(X) and target(Y) ###
X = new_df.drop(columns = 'Class', axis = 1)
Y = new_df['Class']

In [19]:
print(X)

            Time        V1        V2        V3        V4        V5        V6  \
174305  121859.0 -0.346410  0.398823 -2.212968 -2.051168  0.598982 -2.358272   
269283  163614.0  2.056941  0.115745 -1.703518  0.434246  0.369519 -0.902841   
78613    57614.0  1.203190  0.056183  0.576976  0.572712 -0.708321 -0.835329   
71973    54515.0 -0.205346  1.065287  1.166512  1.501682 -0.384694 -0.340569   
170113  120023.0  2.038026 -0.144458 -1.219570  0.202584  0.110969 -0.583490   
...          ...       ...       ...       ...       ...       ...       ...   
279863  169142.0 -1.927883  1.125653 -4.518331  1.749293 -1.566487 -2.010494   
280143  169347.0  1.378559  1.289381 -5.004247  1.411850  0.442581 -1.326536   
280149  169351.0 -0.676143  1.126366 -2.213700  0.468308 -1.120541 -0.003346   
281144  169966.0 -3.113832  0.585864 -5.399730  1.817092 -0.840618 -2.943548   
281674  170348.0  1.991976  0.158476 -2.583441  0.408670  1.151147 -0.096695   

              V7        V8        V9  .

In [20]:
print(Y)

174305    0
269283    0
78613     0
71973     0
170113    0
         ..
279863    1
280143    1
280149    1
281144    1
281674    1
Name: Class, Length: 984, dtype: int64


In [21]:
## Splitting the data into training & testing data ##
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size= 0.2, stratify = Y , random_state= 2)

Here, the train test split is 8:2. 

sratify(Y) based on Y will evenly distribute the data, otherwise the Y - 0 & 1 distribution will be very different in the training and testing data.

random_state = 2 ==> This helps control the random shuffling of the data. This means that if the same 'random_state' value is used in multiple runs of the train-test split function, you will get the same split of data each time.

In [22]:
print(X.shape, X_train.shape, X_test.shape)

(984, 30) (787, 30) (197, 30)


Training the model on X_train & Y_train. Logistic Regression is used here because for binary classification its the best suitable model. Other models can be used too but accuracy will be different.

In [23]:
### Training the model ###

model = LogisticRegression() ## creating Logistic regression object

model.fit(X_train, Y_train)  ## Training the model on training data


LogisticRegression()

In [25]:
### Model Evaluation based on accuracy score ###

## Accuracy on training data

X_train_prediction = model.predict(X_train) ## the predicted values for X_train will get stored here

Accuracy_training_data = accuracy_score(X_train_prediction, Y_train) ##predicted value will get compared with Y_train(original values)

In [26]:
print('Training data accuracy:', Accuracy_training_data)

Training data accuracy: 0.9161372299872935


In [28]:
## Accuracy on test data ##

X_test_prediction = model.predict(X_test) ## the predicted values for X_test will get stored here

Accuracy_test_data = accuracy_score(X_test_prediction, Y_test) ## comparison b/w Y_test and X_test_prediction (predicted values)

print('Test data accuracy:', Accuracy_test_data)

Test data accuracy: 0.9137055837563451


Here, this model is perfect and the accuracy score of training and testing data is almost same.

Overfitting - If Train_accuracy_score > Test_accuracy_score --- Where model gets overtrained on training data

Underfitting - If Train_accuracy_score < Test_accuracy_score --- Where the model gets underfit data.
